File: 102.py
#102. Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def levelOrder(root):
    if(root is None):
        return []
    q, ret = [root], []
    while(len(q) != 0):
        size, curr = len(q), []
        for i in range(size): 
            n = q.pop(0)
            if(n.left):
                q.append(n.left)
            if(n.right):
                q.append(n.right)
            curr.append(n.val)
        ret.append(curr)
    return ret

#testcase : root = [3,9,20,null,15,7]
a = TreeNode(3, TreeNode(9), TreeNode(20))
a.right.left = TreeNode(15)
a.right.right = TreeNode(7)

print(levelOrder(a))

File: golang/102.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	d := TreeNode{Val: 7}
	c := TreeNode{Val: 15}
	b := TreeNode{Val: 20, Left: &c, Right: &d}
	a := TreeNode{Val: 9}
	head := TreeNode{Val: 3, Left: &a, Right: &b}

	fmt.Println(levelOrder(&head))
}

// DESC: Binary Tree Level Order Traversal
// Implementing the Breadth First Search using Queue
func levelOrder(root *TreeNode) [][]int {
	q := []*TreeNode{}
	res := [][]int{}
	if root == nil {
		return res
	}
	q = append(q, root)
	for len(q) != 0 {
		n := len(q)
		list := []int{}
		for i := 0; i < n; i++ {
			curr := q[0]
			if curr.Left != nil {
				//Enqueue
				q = append(q, curr.Left)
			}
			if curr.Right != nil {
				//Enqueue
				q = append(q, curr.Right)
			}
			//Dequeue
			list = append(list, curr.Val)
			q = q[1:]
		}
		res = append(res, list)
	}
	return res
}
